### PR TITLE
validation perf improvements

### PR DESCRIPTION
**Pull request by jbellenger**
Howdy from Twitter! We're in the final stages of migrating our graphql api to graphql-java. We're _very_ excited about the performance and stability we've measured in graphql-java as well as the healthy state of this project and it's broad adoption in the industry.

We've seen that for some of our larger queries we're spending over 5s in query validation. It looks like most of that time is caused by RulesVisitor traversing from fragment spreads into fragment definitions and that that it will revisit fragments that it has already checked. I suspect this has ~exponential complexity for worst-case fragment graphs.

We humbly propose this PR to improve RulesVisitor performance. This changes it to roughly linear complexity and shows a 5000% speedup when validating our large queries. I've included a jmh benchmark which shows a ~1800% speedup on one the existing large schema/query resources:

```
ValidatorBenchmark.oldRuleVisitor  avgt   30  209.684 ± 23.927  ms/op
ValidatorBenchmark.newRuleVisitor  avgt   30   38.141 ±  4.131  ms/op
```

This diff is slightly noisy. The crux of it is:
  1. in `checkFragmentSpread`, don't traverse into fragment definitions that have already been checked
  1. doing this requires keeping a little more state around. It helps to reuse the existing RulesVisitor rather than creating a new one for each FragmentDefinition traversal
  1. when working with only one `RulesVisitor`, it helps to be able to restore state from previous scopes. This diff goes about this by wrapping some of the properties in a `java.util.Stack`

I'm open to literally any/all feedback here. Thanks gang.

### CODE REVIEW

File: src/test/groovy/graphql/validation/rules/NoUnusedVariablesTest.groovy [hunk 1]
```diff
@@ -89,8 +85,33 @@ class NoUnusedVariablesTest extends Specification {
...
+            ...FragA
+        }
+        """
+
+        when:
+        traverse(query)
 
+        then:
+        errorCollector.containsValidationError(ValidationErrorType.UnusedVariable)
+    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/validation/rules/NoUnusedVariablesTest.groovy:116]: were we missing a test for this case?

**Review by bbakerman — APPROVED** [review]
Thank you for this PR.  The fact that no tests had to be changed and we have a benchmark run for this gives me great confidence that it will do what is promised

One interesting thing we have lesrnt about graphql is that queries are more pathological then we think.

More and more front ends are "generating" their queries via "fragment decls" and hence computers end up writing horrendously complicated queries.

And hence parsing / validation has over time become more of a contention point= than it used to be.

File: src/main/java/graphql/validation/RulesVisitor.java [hunk 2]
```diff
@@ -25,41 +24,33 @@
...
+@SuppressWarnings("rawtypes")
 public class RulesVisitor implements DocumentVisitor {
-
-    private final ImmutableList<AbstractRule> rules;
     private final ValidationContext validationContext;
-    private boolean subVisitor;
-    private final List<AbstractRule> rulesVisitingFragmentSpreads = new ArrayList<>();
-    private final Map<Node, List<AbstractRule>> rulesToSkipByUntilNode = new IdentityHashMap<>();
-    private final Set<AbstractRule> rulesToSkip = new LinkedHashSet<>();
+    private int fragmentSpreadVisitDepth;
```
Comment (jbellenger) [inline src/main/java/graphql/validation/RulesVisitor.java:30]: this does the same job as `subVisitor`

File: src/main/java/graphql/validation/RulesVisitor.java [hunk 3]
```diff
@@ -25,41 +24,33 @@
...
+    private final Stack<Node> popRulesStackOnLeave = new Stack<>();
```
Comment (jbellenger) [inline src/main/java/graphql/validation/RulesVisitor.java:31]: this does the same job as `rulesToSkipByUntilNode`

File: src/main/java/graphql/validation/RulesVisitor.java [hunk 4]
```diff
@@ -25,41 +24,33 @@
...
+    private final Stack<ImmutableList<AbstractRule>> rulesStack = new Stack<>();
```
Comment (jbellenger) [inline src/main/java/graphql/validation/RulesVisitor.java:32]: this does a similar job as `rulesToSkip`. Whereas rulesToSkip describes muted rules that should be removed from iteration, `rulesStack` describes the subset of rules that should be iterated. Rules are muted by pushing a new list on the stack that does not contain the muted rules.

File: src/test/groovy/graphql/validation/rules/NoUnusedVariablesTest.groovy (hunk 1 above)
Comment (jbellenger) [reply src/test/groovy/graphql/validation/rules/NoUnusedVariablesTest.groovy:116]: I didn't see one; the `variables not used` test in this file is close but doesn't exercise fragments, which seems like a potential blindspot.

**Comment by andimarek** [comment]
high level comment: I didn't have time to look really deeper: we are very happy with these kind of improvements in general. Performance is one of our main focus areas.

One thing: I don't really trust our validation tests as much as in other parts in our code base. 

One specific concern: if we have fragments which are used multiple times, will we miss something?
For example: cycle detection. If we have a fragment which causes a cycle, but it is used multiple times. Could that happen? (I am not sure, hypothetical example)

**Review by andimarek — CHANGES_REQUESTED** [review]
Thanks a lot for this work and sorry for the long delay of my review.

We have 3 rules which want to follow fragments spreads when traversing the document:
- No Undefined variables
- No unused variables
- Variables types match rule

As far as I can tell the only reason they need to follow fragment spreads is because they need to consider the current operation: one fragment can be reused across multiple operations in a document. So each of these rules need to check every variable usage in respect to each Operation it is used in.

But this also mean that for every Operation we only need to visit each Fragment definition once, right?

The code seems overall complex for that.
Can we simplify it by making simply sure we don't visit the same Fragment Definition multiple times per Operation? What do you think?

File: src/main/java/graphql/validation/RulesVisitor.java [hunk 5]
```diff
@@ -88,139 +79,108 @@ public void enter(Node node, List<Node> ancestors) {
...
+        ImmutableList<AbstractRule> rulesVisitingFragmentSpreads = filterRulesVisitingFragmentSpreads(rules, true);
         if (rulesVisitingFragmentSpreads.size() > 0) {
-            FragmentDefinition fragment = validationContext.getFragment(fragmentSpread.getName());
-            if (fragment != null && !ancestors.contains(fragment)) {
-                new LanguageTraversal(ancestors).traverse(fragment, new RulesVisitor(validationContext, rulesVisitingFragmentSpreads, true));
+            FragmentDefinition fragment = validationContext.getFragment(node.getName());
+            if (fragment != null && !ancestors.contains(fragment) && !visitedFragmentSpreads.contains(node.getName())) {
+                // Manually traverse into the FragmentDefinition
+                visitedFragmentSpreads.add(node.getName());
+                rulesStack.push(rulesVisitingFragmentSpreads);
```
Comment (andimarek) [inline src/main/java/graphql/validation/RulesVisitor.java:118]: just as an example why I think the current code is too complicated: this is always the same set of 3 rules, we don't need to maintain an extra stack for that. 

Btw: the old code before your PR was not better, nothing todo with your PR directly :-)

File: src/main/java/graphql/validation/RulesVisitor.java [hunk 6]
```diff
@@ -88,139 +79,108 @@ public void enter(Node node, List<Node> ancestors) {
...
-
-
-        for (AbstractRule rule : rules) {
-            if (!subVisitor && (rule.isVisitFragmentSpreads())) continue;
-            rule.checkFragmentDefinition(fragmentDefinition);
+        // If we've encountered a FragmentDefinition and we got here without coming through a
+        // FragmentSpread, then suspend all isVisitFragmentSpread rules for this subtree.
+        // Expect these rules to be checked when when the FragmentSpread is traversed
+        if (fragmentSpreadVisitDepth == 0) {
+            scopeRules = filterRulesVisitingFragmentSpreads(rules, false);
```
Comment (andimarek) [inline src/main/java/graphql/validation/RulesVisitor.java:134]: so these are basically all rules which don't want to visit fragment spreads, meaning all Rules except 3.

**Comment by jbellenger** [comment]
Thanks for the feedback @andimarek!

Agreed that RulesVisitor could be a bit simpler. I've reworked that class to not use a stack internally, though I found it necessary to introduce some new mutable state that was hard to avoid. Overall it feels like this attempt for simplicity may have been a wash but I'd appreciate your feedback here.

> Looking more into it the whole stacks we maintain are only ever needed for Fragment Definition in/out and we actually only ever have two different sets of rules: the ones following fragment spreads and the ones who don't.

This was spot on.

**Comment by andimarek** [comment]
Thanks a lot @jbellenger.

**Pull request merged**